aeneas/src/x86/X86MacroAssembler.v3

```diff
@@ -100,6 +151,44 @@ class X86MacroAssembler extends X86Assembler {
 		mach.fail(Strings.format2("%1: %2", msg, regSet.identify(loc)));
 		return X86MachRegs.SCRATCH;
 	}
+	// convert a location into an SSE register
+	def loc_s(frame: MachFrame, loc: int) -> SSEReg {
+		match (loc) {
+			X86MachRegs.XMM0 => return xmm0;
+			X86MachRegs.XMM1 => return xmm1;
+			X86MachRegs.XMM2 => return xmm2;
+			X86MachRegs.XMM3 => return xmm3;
+			X86MachRegs.XMM4 => return xmm4;
+			X86MachRegs.XMM5 => return xmm5;
+			X86MachRegs.XMM6 => return xmm6;
+			X86MachRegs.XMM7 => return xmm7;
+		}
+		return failSSELocation("required SSE register", loc, frame.conv.regSet);
+	}
+	// convert a location into an SSE register/memory reference
+	def loc_sm(frame: MachFrame, loc: int) -> SSERm {
+		match (loc) {
+			0 => return failSSELocation("unassigned location", loc, frame.conv.regSet);
+			X86MachRegs.XMM0 => return xmm0;
+			X86MachRegs.XMM1 => return xmm1;
+			X86MachRegs.XMM2 => return xmm2;
+			X86MachRegs.XMM3 => return xmm3;
+			X86MachRegs.XMM4 => return xmm4;
+			X86MachRegs.XMM5 => return xmm5;
+			X86MachRegs.XMM6 => return xmm6;
+			X86MachRegs.XMM7 => return xmm7;
+		}
+		var regSet = frame.conv.regSet, wordSize = mach.data.addressSize, offset: int;
+		if (loc >= regSet.calleeStart) offset = wordSize * (loc - regSet.calleeStart);
```

Comment:
Is it possible to factor out the offset calculation, since it doesn't depend on the return value being an SSERm or X86Rm?

Reply:
Maybe? The issue is the `else return failLocation...`. How do we nicely implement that. I'm thinking we could internally output `(int, bool)` for calculating the offset with the `bool` representing if the location has been found and the offset calculated?

Reply:
Done.